### PR TITLE
Try to query the Domain objects first in middleware. Then fall back to Header & DNS.

### PR DESCRIPTION
**Pull request by ericholscher**
It looks like we're actually using fully qualified URL's (http, https, paths) in the Domain modeling, which was there to support subprojects & translations. However, I think we need to just be storing the actual Domain as unique, and using subproject & translation modeling to create appropriate domains for those projects. 

This adds that, and switches to using proper Domain objects on CNAME resolution as the first choice. This makes it so people can use CDN's and other things that break CNAME's currently.

### CODE REVIEW

File: readthedocs/core/validators.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                raise
+            if not value:
+                raise
+            try:
+                idnavalue = value.encode('idna')
+            except UnicodeError:
+                raise exc
+            super(DomainNameValidator, self).__call__(idnavalue)
+
+validate_domain_name = DomainNameValidator()
```
Comment (agjohnson) [inline readthedocs/core/validators.py]: We should open an issue to track this merging into django as well

File: readthedocs/templates/projects/domain_list.html [hunk 2]
```diff
@@ -23,7 +23,7 @@
     {% for domain in object_list %}
         <li {% if domain.machine %}class="help_text"{% endif %}>
         {% if domain.canonical %}<strong>{% endif %}
-        {{ domain.url }}
+        {{ domain.domain }}
```
Comment (agjohnson) [inline readthedocs/templates/projects/domain_list.html]: Might as well define this as a more semantic css selector, ie - `domain-selected`. Specific formatting should be kept out of the html and push to css.

**Comment by agjohnson** [comment]
LGTM, though I noted a change to the html template that wasn't part of the PR here. It's just a small change though

File: readthedocs/core/validators.py (hunk 1 above)
Comment (ericholscher) [inline readthedocs/core/validators.py]: #1707

**Comment by ericholscher** [comment]
What change in the template?

**Pull request merged**